debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/structureformat/extensions/J9ObjectStructureFormatter.java

```diff
@@ -306,55 +379,142 @@ private void printJ9ObjectFields(PrintStream out, int tabLevel, J9ClassPointer l
 		String fieldSignature = J9UTF8Helper.stringValue(fieldShape.nameAndSignature().signature());
-		
+		boolean fieldIsFlattened = ValueTypeHelper.isFieldInClassFlattened(localClazz, fieldName);
 		U8Pointer valuePtr = dataStart;
 		valuePtr = valuePtr.add(fieldOffset);
 
 		padding(out, tabLevel);
 		out.print(String.format("%s %s = ", fieldSignature, fieldName));
-		
+
 		if (fieldShape.modifiers().anyBitsIn(J9FieldSizeDouble)) {
 			out.print(U64Pointer.cast(valuePtr).at(0).getHexValue());
 		} else if (fieldShape.modifiers().anyBitsIn(J9FieldFlagObject)) {
-			AbstractPointer ptr = J9BuildFlags.gc_compressedPointers ? U32Pointer.cast(valuePtr) : UDATAPointer.cast(valuePtr);
-			out.print(String.format("!fj9object 0x%x", ptr.at(0).longValue()));
+			if (fieldIsFlattened) {
+				String heirarchy = "";
+				
+				if (containerIsFlatObject) {
+					for (int i = 0; i < nestingHeirarchy.length; i++) {
+						heirarchy += nestingHeirarchy[i] + ",";
+					}
+				}
+
+				J9ObjectPointer object = J9ObjectPointer.cast(address);
+				out.print(String.format("!j9object 0x%x %s", object.getAddress(), heirarchy + fieldName));
+			} else {
+				AbstractPointer ptr = J9BuildFlags.gc_compressedPointers ? U32Pointer.cast(valuePtr) : UDATAPointer.cast(valuePtr);
+				out.print(String.format("!fj9object 0x%x", ptr.at(0).longValue()));
+			}
 		} else {
 			out.print(I32Pointer.cast(valuePtr).at(0).getHexValue());
 		}
+		if (fieldIsFlattened) {
+			out.print(String.format(" (offset=%d) (%s)", fieldOffset.longValue(), fieldSignature.substring(1, fieldSignature.length() - 1)));
+		} else {
+			out.print(String.format(" (offset=%d) (%s)", fieldOffset.longValue(), className));
+		}
+	
+		if (isHiddenField) {
+			out.print(" <hidden>");
+		}
+	}
+
+	public static void printNestedObjectField(PrintStream out, int tabLevel, J9ClassPointer localClazz, U8Pointer dataStart, J9ClassPointer fromClass, J9ObjectFieldOffset objectFieldOffset, long address, String[] nestingHeirarchy, J9ClassPointer containerClazz) throws CorruptDataException 
+	{
+		J9ROMFieldShapePointer fieldShape = objectFieldOffset.getField();
+		UDATA fieldOffset = objectFieldOffset.getOffsetOrAddress();
+		boolean isHiddenField = objectFieldOffset.isHidden();
+		String fieldName = J9UTF8Helper.stringValue(fieldShape.nameAndSignature().name());
+		String fieldSignature = J9UTF8Helper.stringValue(fieldShape.nameAndSignature().signature());
+		boolean fieldIsFlattened = ValueTypeHelper.isFieldInClassFlattened(localClazz, fieldName);
+		U8Pointer valuePtr = dataStart;
+		valuePtr = valuePtr.add(fieldOffset);
 		
-		out.print(String.format(" (offset=%d) (%s)", fieldOffset.longValue(), className));
+		if (fieldIsFlattened) {
+			padding(out, tabLevel);
+			out.println(String.format("%s %s { // (EA=%s) (offset=%d)", fieldSignature.substring(1, fieldSignature.length() - 1), fieldName, valuePtr.getHexAddress(), fieldOffset.longValue()));
```

Comment:
We should be consistent about whether `EA=...` is enclosed in parentheses (it's not above).